### PR TITLE
nixos/mysql: run postStart as an unprivileged user

### PR DESCRIPTION
**Pull request by aanderse**
###### Motivation for this change
- general module cleanup
- remove the need to run any part of this module as `root`
- keep consistent with `upstream`
- keep consistent with NixOS best practices

My intention is to squash e3c210dfd19f60587a98c055266570d0f950edec (`nixos/mysql: run ExecStartPost as an unprivileged user`) and a0539fbe8635205ce3e5d0bf4daba316a92651ba (`nixos/mysql: move ExecStartPost into postStart`) before merge.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

ping @thorstenweber83

### CODE REVIEW

**Comment by aanderse** [comment]
@GrahamcOfBorg test mysql

**Review by Izorkin — APPROVED** [review]

**Review by florianjacob — APPROVED** [review]
From a read-only review, this looks like a well-done and necessary cleanup of the module.

File: nixos/modules/services/databases/mysql.nix [hunk 1]
```diff
@@ -357,125 +365,127 @@ in
...
-                        ) | ${mysql}/bin/mysql -u root -N
+                echo "MySQL daemon not yet started. Waiting for 1 second..."
+                count=$((count++))
+                sleep 1
+            done
+          ''}
+
+          if [ -f ${cfg.dataDir}/mysql_init ]
+          then
+              # While MariaDB comes with a 'mysql' super user account since 10.4.x MySQL does not
```
Comment (florianjacob) [inline nixos/modules/services/databases/mysql.nix:401]: While …, MySQL does not. (missing comma)

Comment (aanderse) [reply nixos/modules/services/databases/mysql.nix]: Thanks for the review and suggestion.

File: nixos/modules/services/databases/mysql.nix [hunk 2]
```diff
@@ -357,125 +365,127 @@ in
...
+        postStart = let
+          # The super user account to use on *first* run of MySQL server
+          superUser = if isMariaDB then cfg.user else "root";
+        in ''
+          ${optionalString (!hasNotify) ''
+            # Wait until the MySQL server is available for use
+            count=0
+            while [ ! -e /run/mysqld/mysqld.sock ]
+            do
+                if [ $count -eq 30 ]
```
Comment (aszlig) [inline nixos/modules/services/databases/mysql.nix:387]: I know, that this has been like this before, but I think 30 seconds might not be quite enough, especially for large instances. I've got two servers in production which need several minutes to start up (they're using a custom MySQL NixOS module). They're MariaDB instances however and thus have support for notify, but longer startup times (especially if eg. innodb tables need to be recovered) could happen for MySQL as well.

**Review by aszlig — APPROVED** [review]

File: nixos/modules/services/databases/mysql.nix (hunk 2 above)
Comment (aanderse) [reply nixos/modules/services/databases/mysql.nix:387]: @aszlig if you want to provide a value for me I'd be happy to add a commit for you in my next PR against this module (which is coming soon). I've seen 600 seconds as a timeout on some other distros for `mysql`.

**Comment by aanderse** [comment]
Merging based on review and feedback. Thanks all :tada:

**Comment by aanderse** [comment]
@GrahamcOfBorg test mysql

**Comment by aanderse** [comment]
The `aarch64-linux` variant of the test has been failing for a while... so not this PR :man_shrugging:

**Pull request merged**